subprojects/model-core/src/main/java/org/gradle/internal/instantiation/AbstractClassGenerator.java

```diff
@@ -883,9 +890,12 @@ void applyTo(ClassInspectionVisitor visitor) {
             if (!hasFields) {
                 visitor.mixInManaged();
             }
-            if (!readOnlyProperties.isEmpty()) {
+            if (!readOnlyProperties.isEmpty() || !readOnlyNestedProperties.isEmpty()) {
                 visitor.mixInServiceInjection();
```

Comment:
If we have read only properties, then services need to be mixed in so that `ObjectFactory` can be obtained from the services to instantiate the values of the fields.